### PR TITLE
Switch to DE in NGOpt8 if multiobjective

### PR DESCRIPTION
**Pull request by jrapin**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

Replaces #915 

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](https://facebookresearch.github.io/nevergrad/contributing.html) document and completed the CLA (see [**CLA**](https://facebookresearch.github.io/nevergrad/contributing.html#contributor-license-agreement-cla)).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: nevergrad/optimization/test_optimizerlib.py [hunk 1]
```diff
@@ -407,29 +407,30 @@ def test_parallel_es() -> None:
...
-    (2, 1, 8., 40, False),
-    #(2, 3, 8., 70, True),
-    (5, 1, 1., 150, False),
-    #(5, 3, 1., 225, True),
-    (5, 1, 8., 150, False),
-    #(5, 3, 8., 500, True),
-    (8, 27, 8., 200, True),
-    (9, 27, 8., 700, True),
-    #(10, 27, 8., 400, False),
+        # (2, 3, 1., 120, False),
```
Comment (jrapin) [inline nevergrad/optimization/test_optimizerlib.py:410]: @teytaud I deactivated this test which was flaky, I am a bit concerned that all of them may be. If so they must be improved or removed

File: nevergrad/optimization/optimizerlib.py [hunk 2]
```diff
@@ -2131,6 +2131,12 @@ def _select_optimizer_cls(self) -> base.OptCls:
 
         return optimClass
 
+    def _num_objectives_callback(self) -> None:
+        if self.num_objectives > 1:
+            if self.noise_from_instrumentation or not self.has_noise:
+                # override at runtime
+                self._optim = DE(self.parametrization, self.budget, self.num_workers)
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:2138]: isnt TwoPointsDE better most of the times?

File: nevergrad/optimization/optimizerlib.py [hunk 3]
```diff
@@ -2131,6 +2131,12 @@ def _select_optimizer_cls(self) -> base.OptCls:
 
         return optimClass
 
+    def _num_objectives_callback(self) -> None:
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:2134]: this will happen when we get the first tell.

File: nevergrad/parametrization/utils.py [hunk 4]
```diff
@@ -24,8 +24,6 @@ def __init__(
         self,
         deterministic: bool = True,
         deterministic_function: bool = True,
-        monoobjective: bool = True,
-        not_manyobjective: bool = True,
```
Comment (jrapin) [inline nevergrad/parametrization/utils.py:28]: no more needed

File: nevergrad/optimization/optimizerlib.py (hunk 2 above)
Comment (teytaud) [reply nevergrad/optimization/optimizerlib.py:2138]: For the multiobjective setting you know better than me...

File: nevergrad/optimization/test_optimizerlib.py (hunk 1 above)
Comment (teytaud) [reply nevergrad/optimization/test_optimizerlib.py:410]: Solved in another PR.

**Comment by teytaud** [comment]
I'm a bit scared of the runtime overriding of the optimizer.
And in some cases maybe we have the knowledge "multiobjectiveness" a priori and this should be used.

**Review by teytaud — APPROVED** [review]
A bit scared, but I can't point out any problem :-)

File: nevergrad/optimization/optimizerlib.py (hunk 2 above)
Comment (jrapin) [reply nevergrad/optimization/optimizerlib.py:2138]: Well no, all DE variants will have the same selection process anyway, so it comes down to which DE variant we would prefer.

**Comment by jrapin** [comment]
> I'm a bit scared of the runtime overriding of the optimizer.

Why?

> And in some cases maybe we have the knowledge "multiobjectiveness" a priori and this should be used.

for this you have several option:
- currently official: `optimizer.tell(ng.p.MultiobjectiveReference(), [5, 5])`  that provides a reference point for volumes, and at the same time the number of objectives before the first ask
- undocumented by now available through this PR: `optimizer.num_objectives = n`. This will trigger the selection for NGOpt if need be. If this is not coherent with previous (or future) number of objectives, it raises an error, so safe

**Comment by teytaud** [comment]
> > I'm a bit scared of the runtime overriding of the optimizer.
> 
> Why?
> 
> > And in some cases maybe we have the knowledge "multiobjectiveness" a priori and this should be used.
> 
> for this you have several option:
> 
> * currently official: `optimizer.tell(ng.p.MultiobjectiveReference(), [5, 5])`  that provides a reference point for volumes, and at the same time the number of objectives before the first ask
> * undocumented by now available through this PR: `optimizer.num_objectives = n`. This will trigger the selection for NGOpt if need be. If this is not coherent with previous (or future) number of objectives, it raises an error, so safe

Convincing. Cool :-)

**Comment by teytaud** [comment]
... and IIUC a cool thing is that the multiobjective-nature of an Experiment in which there is a multiobjective reference point is provided is detected early, i.e. before the first ask.

**Comment by jrapin** [comment]
> ... and IIUC a cool thing is that the multiobjective-nature of an Experiment in which there is a multiobjective reference point is provided is detected early, i.e. before the first ask.

Yes. Otherwise the reference point would different from one run to the next, and therefore we would not be able to compare anything (the volumes would need be comparable)

**Pull request merged**